File: RayTracer/Tracer/ViewPlane.cs
using RayTracer.Common;
using System;

namespace RayTracer.Tracer
{
    public class ViewPlane
    {
        //TO DO: make NOT singleton

        public static ViewPlane Instance;
        float worldWidth;
        float worldHeight;
        int pixelHeight;
        int pixelWidth;
        Point3 position;
        Point3 upperLeft;
        Point3 unitRight;
        Point3 unitDown;

        public int PixelWidth
        {
            get { return pixelWidth; }
            set { pixelWidth = value; }
        }

        public int PixelHeight
        {
            get { return pixelHeight; }
            set { pixelHeight = value; }
        }


        public ViewPlane(int width, int height)
        {
            pixelWidth = width;
            pixelHeight = height;


            worldHeight = 2.0f * (float)Math.Tan((Camera.Instance.FieldOfView / 2f) * (float)Math.PI / 180.0f);
            worldWidth = worldHeight * (pixelWidth / (float)PixelHeight);

            PreCalculate();

        }


        void PreCalculate()
        {
            upperLeft = GetUpperLeft();
            unitRight = (Camera.Instance.U * (worldWidth / (float)PixelWidth) * -1);
            unitDown = (Camera.Instance.V * (worldHeight / (float)PixelHeight) * -1);
        }

        Point3 GetUpperLeft()
        {

            Point3 center = Camera.Instance.Position;
            center += Camera.Instance.W;
            position = center;
            Point3 upperLeft =
                center
                + Camera.Instance.U * (worldWidth / 2.0f)    //U is left
                + Camera.Instance.V * (worldHeight / 2.0f);  //V is up


            return upperLeft;
        }

        public Point3 GetNewLocation(int col, int row)
        {
            Point3 newLocation =
                upperLeft
                + unitRight * (col + .5f)
                + unitDown * (row + .5f);
            return newLocation;
        }

        public Point3 GetNewLocation(int col, int row, float colOffset, float rowOffset)
        {
            Point3 newLocation =
                upperLeft
                + unitRight * (col + .5f + colOffset)
                + unitDown * (row + .5f + rowOffset);
            return newLocation;
        }
    }
}

File: RayTracer - BVH - CUDA/RayTracerCUDA/ViewPlane.cpp
#include "ViewPlane.h"
//#define _USE_MATH_DEFINES
//#include <cmath>  
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

bool ViewPlane::flag = false;
ViewPlane * ViewPlane::instance = nullptr;

void ViewPlane::Init(int width, int height)
{
	pixelH = height;
	pixelW = width;
	worldH = 2.f * tanf((Camera::Instance()->fov / 2.f) * (float)M_PI / 180.f);
	worldW = worldH * (pixelW / (float)pixelH);

	Precalculate();

}

void ViewPlane::Precalculate()
{
	upperLeft = getUpperLeft();
	unitRight = ((Camera::Instance()->U * (worldW / (float)pixelW)) * -1.f);
	unitDown = ((Camera::Instance()->V * (worldH / (float)pixelH)) * -1.f);
}

vec3 & ViewPlane::getUpperLeft()
{
	vec3 c = Camera::Instance()->pos;
	c = c + (Camera::Instance()->W);
	pos = c;
	vec3 res =
		c
		+ (Camera::Instance()->U * (worldW * .5f))
		+ (Camera::Instance()->V * (worldH * .5f));
	res[3] = 1;
	return res;
}

vec3 ViewPlane::getNewLocation(int col, int row)
{
	vec3 uleft = upperLeft;
	vec3 uRight = unitRight;
	vec3 uDown = unitDown;
	vec3 res =
		uleft +
		uRight * (col + .5f)
		+ uDown * (row + .5f)
		;
	res[3] = 1;
	return res;

}

ViewPlane::ViewPlane()
{
	Init(1, 1);
}

ViewPlane::ViewPlane(const ViewPlane &)
{
}
//
//ViewPlane & ViewPlane::operator=(const ViewPlane &)
//{
//	return Instance();
//}


ViewPlane::~ViewPlane()
{
}
